README.md
# couch-statsd-backend

Backend for [StatsD](https://github.com/etsy/statsd), which
publishes stats to [CouchDB](http://couchdb.apache.org/)

## Installation

Within your statsd install:

    npm install couch-statsd-backend


## Configuration


	{

	    port: 8125

	    ,deleteIdleStats: true

	    //, couchhost: '192.168.3.21'    //default: localhost
	    //, couchport: '5984'            //default: 5984
	    , couchdb: 'abtest'


	    //all below are optional
	    //, debug: true

	    //,bulk_size: 2500      // (0 to disable)
	    //,bulk_wait: 50

	    //, id_prefix: "c0-"
	    //, id_prefix: process.pid

	    //choose just one
	    //, id_generator: "couch"
	    //, id_generator: "cuid"
	    //, id_generator: "node-uuid-v1"
	    //, id_generator: "node-uuid-v4"
	    //, id_generator: "custom"
	    //only needed for 'custom'
	    //, uuid_generator: function(num){ return Math.round(new Date().getTime() / 1000) + '-' + num; }

	      , backends: [ "./backends/couch" ]

	}


### Doc _id notes

id_prefix: Depending on the setup it may be useful to prefix the _id with a string to identify the 
statsd process which sent the doc.  Default is not to add anything.


It is possible to choose how the couch doc._id is generated - by default a single request to couchhost/_uuids 
is made per flush but an _id can be generated with javascript instead. 

If 'cuid' or 'node-uuid' are chosen you may need to first install the npms as they are not installed by default.

    npm install cuid

and

    npm install node-uuid


If custom is chosen then uuid_generator must also be set eg:

      , id_generator: "custom"
      , uuid_generator: function(num){ return Math.round(new Date().getTime() / 1000) + '-' + num; }  

Note: The uuid_generator function is passed 'num' which is incremented each 
call within a flush, this is reset on each flush.


###Bulking

bulk_size: Number of docs to submit per bulk request. To disable breaking up of bulk doc send set this to 0 (if disabled all docs will be sent in a single request).

bulk_wait: Time to wait (ms) between sending each bulk of docs for a flush to couch.
Be careful with this if total time to insert all docs per flush > time period of flush then 
docs to be posted will just keep growing.  

It should be ok for the total time to insert all the bulks to run over the statsd flush period 
if a sudden rush of metrics arrive once in a while but if this keeps happening i suspect 
node will run out of memory to hold the backlog. 


##Supported Metrics

* Counters
* Counter Rates
* Gauges
* Timers
* Timer Data
* Sets

###Counters & Counter Rates

Counters and counter rates are combined into one doc - this can be easily 
modified to create a separate docs for the counter_rates by un-commenting 
the a few lines.

    echo "myservice.mycount:67|c" | nc -n -w 1 -u 192.168.3.22 8125 

Will add a doc like:

    {
       "_id": "f83a62b3456f4c2451cb7a166100f3e8",
       "_rev": "1-d68546bae9169fa4b4a72b3ffe4167b9",
       "type": "counter",
       "name": "myservice.mycount",
       "count": 67,
       "rate": 6.7,
       "ts": 1424818189
    }


If two counts of the same name within same flush period

    echo "myservice.mycount:20|c" | nc -n -w 1 -u 192.168.3.22 8125 
    echo "myservice.mycount:30|c" | nc -n -w 1 -u 192.168.3.22 8125 

Will add a doc like:

    {
       "_id": "f83a62b3456f4c2451cb7a1661010a10",
       "_rev": "1-89dc6493c2418acb351e9767e6653c65",
       "type": "counter",
       "name": "myservice.mycount",
       "count": 50,
       "rate": 5,
       "ts": 1424818219
    }


###Gauges

    echo "myservice.mygauge:42|g" | nc -n -w 1 -u 192.168.3.22 8125 

Will add a doc like:

    {
       "_id": "f83a62b3456f4c2451cb7a1661012b57",
       "_rev": "1-390bc981b63e2ca16d2ad42c042f9c1b",
       "type": "gauge",
       "name": "myservice.mygauge",
       "value": 42,
       "ts": 1424819100
    }

###Timers & Timer Data

Timers and timer data are combined into one doc - this can be easily 
modified to create a separate docs for the timer_data by un-commenting 
the a few lines.


    echo "myservice.mytimer:231|ms" | nc -n -w 1 -u 192.168.3.22 8125

Will add a doc like:

    {
       "_id": "2d0f912baefa5e8c6e131e29c4482e67",
       "_rev": "1-30071b20880b8f0139a403856e0365d9",
       "type": "timers",
       "name": "myservice.mytimer",
       "durations": [
           231
       ],
       "mean_90": 231,
       "upper_90": 231,
       "sum_90": 231,
       "std": 0,
       "upper": 231,
       "lower": 231,
       "count": 1,
       "count_ps": 0.1,
       "sum": 231,
       "mean": 231,
       "median": 231,
       "ts": 1425055422
    }
    

If more than one timer is received withing the same flush interval eg:    

    echo "myservice.mytimer:30|ms" | nc -n -w 1 -u 192.168.3.22 8125
    echo "myservice.mytimer:23|ms" | nc -n -w 1 -u 192.168.3.22 8125
    echo "myservice.mytimer:51|ms" | nc -n -w 1 -u 192.168.3.22 8125

Will add a doc like:

    {
       "_id": "2d0f912baefa5e8c6e131e29c44835d1",
       "_rev": "1-5786babae3d5c82d6511a5ee7770fd86",
       "type": "timers",
       "name": "myservice.mytimer",
       "durations": [
           23,
           30,
           51
       ],
       "mean_90": 34.666666666666664,
       "upper_90": 51,
       "sum_90": 104,
       "std": 11.897712198383164,
       "upper": 51,
       "lower": 23,
       "count": 3,
       "count_ps": 0.3,
       "sum": 104,
       "mean": 34.666666666666664,
       "median": 30,
       "ts": 1425055602
    }



###Sets 

    echo "myservice.myset:129|s" | nc -n -w 1 -u 192.168.3.22 8125 

Will add a doc like:

    {
       "_id": "f83a62b3456f4c2451cb7a16617f1d84",
       "_rev": "1-07926b51ef9b5a163a9bde8f3635223c",
       "type": "set",
       "name": "myservice.myset",
       "set": {
           "129": "129"
       },
       "ts": 1424819898
    }

*I am not sure if the set data structure is correct?


##Notes/Future Changes
A test db with snappy compression of 2 million docs consumed 0.5Gb of space.  

I suspect having the full field names in each document increases the storage requirements so may be better to save the metric field names as single letters for example for a 'counter' type:


    {
       "_id": "f83a62b3456f4c2451cb7a1661010a10",
       "_rev": "1-89dc6493c2418acb351e9767e6653c65",
       "t": "c",
       "n": "myservice.mycount",
       "c": 50,
       "r": 5,
       "ts": 1424818219
    }

where:

    t = type  (where c = counter, g = gauge, t = timer and s = set)
    n = name
    c = count
    r = rate
    
It may even be worth encoding the 'type' within the '_id' ?


Also on testing i did hit issues when adding more that 20k documents every 3 seconds (before i added the bulk stuff so need to retest)
- not 100% sure where the bottle neck was.
Look at sending each bulk batch off to a different couchdb server in a round robin fasion so each couchdb doesnt need to handle all 
metrics from a single flush - this may not be needed with couch 2.0 clustering stuff :). 


CHANGES
0.1.?   

-----
Below published to npm

0.1.5
Minor fix and clearup to console log.

0.1.4
Added timer metrics examples and better config example
Break bulk sends to couch into chunks,  
More config nobs: bulk_size, bulk_wait
Add bulk info to readme
Fix for default id generator to be couch not cuid

0.1.3   
Remove s from gauge type
Config example update
Readme updates

0.1.2	
Add CHANGES 
Fix test for adding rate to counters doc
Clear up sets docs 

0.1.1	
Update to config and readme

0.1.0	
Initial commit.


backends/couch.js


var nano = require('nano');

var flush_stats = function (timestamp, metrics) {

  console.log('Flushing stats at', new Date(timestamp * 1000).toString());
//  console.log(JSON.stringify(metrics.counters['statsd.packets_received']));
//  console.log(JSON.stringify(metrics.counter_rates['statsd.packets_received']));

  var docs = [];

  for(var myname in metrics.counters) {
	if (myname == 'statsd.bad_lines_seen'   && metrics.counters[myname] == '0') continue; 
	if (myname == 'statsd.packets_received' && metrics.counters[myname] == '0') continue; 

	var m = { };
        m['type'] = 'counter';
        m['name'] = myname;
        m['count'] = metrics.counters[myname];
        if( typeof(metrics.counter_rates[myname]) == 'number' ) {
	        m['rate'] = metrics.counter_rates[myname];
	}
        if (debug) console.log(JSON.stringify(m));
 	docs.push(m);
  }

/* un-comment this if you want counter_rate metrics in its own doc and comment the above counter_rate bit
  for(var myname in metrics.counter_rates) {
	if (myname == 'statsd.bad_lines_seen'   && metrics.counter_rates[myname] == '0') continue; 
	if (myname == 'statsd.packets_received' && metrics.counter_rates[myname] == '0') continue; 

	var m = { };
        m['type'] = 'counter_rate';
        m['name'] = myname;
        m['rate'] = metrics.counter_rates[myname];
        if (debug) console.log(JSON.stringify(m));
 	docs.push(m);
  }
*/

  for(var myname in metrics.gauges) {
	if (myname == 'statsd.timestamp_lag' && metrics.gauges[myname] == '0') continue; 
	var m = { };
        m['type'] = 'gauge';
        m['name'] = myname;
        m['value'] = metrics.gauges[myname];
        if (debug) console.log(JSON.stringify(m));
 	docs.push(m);
  }

  for(var myname in metrics.timers) {
	var m = { };
        m['type'] = 'timers';
        m['name'] = myname;
        m['durations'] = metrics.timers[myname];

        //lets also add the timer_data for this timer by this name if it exits to this doc
        //console.log(JSON.stringify(metrics.timer_data[myname]));
        if( typeof(metrics.timer_data[myname]) === 'object' ) {
	        data = metrics.timer_data[myname];
		for(var dataname in data){
			m[dataname] = data[dataname]
		} 
        }
        if (debug) console.log(JSON.stringify(m));
 	docs.push(m);
  }

/* un-comment this if you want timer_data metrics in its own doc and comment the above timer_data bits
  for(var myname in metrics.timer_data) {
	var m = { };
        m['type'] = 'timer_data';
        m['name'] = myname;
        data = metrics.timer_data[myname];
	for(var dataname in data){
		m[dataname] = data[dataname]
	} 
        if (debug) console.log(JSON.stringify(m));
 	docs.push(m);
  }
*/


  for(var myname in metrics.sets) {
	var m = { };
        m['type'] = 'set';
        m['name'] = myname;
        m['set'] = metrics.sets.store[myname];
        if (debug) console.log(JSON.stringify(m));
 	docs.push(m);
  }
  

   if(docs.length > 0){
	if(id_generator == 'couch'){
	     nano.request( { db: "_uuids", path: "?count=" + docs.length }, function(err, uuids) {
	       if((err) || !uuids) { 
			console.log(err); 
               }else{
		  post_docs(docs,uuids.uuids,timestamp,0);
	       }
	     });
     	}else{
            uuids = [];
	    for (var i = 0; i < docs.length; i++) {
	       uuids.push(uuid_generator(i));
	    }
  	    if (debug) console.log('uuids: ' + JSON.stringify(uuids));
	    post_docs(docs, uuids, timestamp, 0);
     	}
     }

     
};


var post_docs = function(docs,uuids,timestamp,chunk_count) {

//	  for(var doc in docs) {
//		docs[doc]._id =  id_prefix + uuids.pop();
//		docs[doc].ts = timestamp;
//	  }

	  var this_bulk_size = bulk_size;
	  if ( (docs.length < bulk_size) || (bulk_size == 0) ) {
		 this_bulk_size = docs.length;
	  }
    
          var chunk = [];        
          var doc = false;
	  var i = 0;
          while(i < this_bulk_size) {
		doc = docs.pop();
		doc._id =  id_prefix + uuids.pop();
		doc.ts = timestamp;
		chunk.push(doc); 
		i++;
          }          

	  if (debug) console.log('docs: ' + JSON.stringify(docs));

	  var bulk = {};
          bulk.docs = chunk;

	  var db = nano.use(couchdb);

          db.bulk(bulk, '', function(err, ret){
             chunk_count++;
	     if((err) || !ret){
	          console.log('err: ' + JSON.stringify(err));    //send metric ?
                  //need to check though error and see if id collision or something else
	     }else{
		  console.log('added chunk: ' + chunk_count + ', ' +  ret.length + ' docs'); 
		  if (debug) console.log('docs: ' + JSON.stringify(ret));
	     }
	  });
	  if(docs.length > 0){
		setTimeout( function(){ post_docs(docs,uuids,timestamp,chunk_count); }, bulk_wait );
	  }

	
};



var status = function(write) {
  ['lastFlush', 'lastException'].forEach(function(key) {
    write(null, 'console', key, this[key]);
  }, this);
};





var id_prefix;
var id_generator;
var debug;
var bulk_size;
var bulk_wait;

var couchdb = false;

exports.init = function(startupTime, config, events) {

  debug = config.debug || false;


  id_prefix = config.id_prefix || '';
  if(id_prefix != '') {
      console.log('id_prefix: ' + id_prefix);
  }

  id_generator = config.id_generator || 'couch';
  switch(id_generator) {
	case 'couch':
            id_generator = 'couch';
            break;
	case 'cuid':
            var cuid = require('cuid');
            uuid_generator = function(){ return cuid(); };
	    break;
	case 'node-uuid-v1':
            var nodeuuid = require('node-uuid');
            uuid_generator = function(){ return nodeuuid.v1(); }
	    break;
	case 'node-uuid-v4':
            var nodeuuid = require('node-uuid');
            uuid_generator = function(){ return nodeuuid.v4(); }
	    break;
        case 'custom':
            //eg: uuid_generator = function(num){ return Math.round(new Date().getTime() / 1000) + '-' + num; }
            if(typeof(config.uuid_generator) != 'function'){
		console.log('custom id generation selected but config.uuid_generator is not set or is not a function');
		return false;
	    }
            uuid_generator = config.uuid_generator;
	    break;
        default:            
	    console.log('config.id_generator:' + id_generator + ' is not known select one of: couch|cuid|node-uuid-v1|node-uuid-v4|custom');
	    return false;
  }
  console.log('using: ' + id_generator + ' for id generation');

  var couchhost = config.couchhost || 'localhost';
  var couchport = config.couchport || '5984';
  
  if (typeof config.couchdb !== 'string') {
    console.log('config.couchdb is not set, config.couchdb must be set.');
    return false;
  }
  couchdb = config.couchdb;

  nano = nano('http://' + couchhost + ':' + couchport);

  if(config.bulk_size == 0) {
          bulk_size = 0;
          console.log('bulking disabled');
  }else{
	  bulk_size = config.bulk_size || 2500;
	  bulk_wait = config.bulk_wait || 50;
          console.log('bulk_size: ' + bulk_size);
          console.log('bulk_wait: ' + bulk_wait);
  }


  events.on('flush', flush_stats);
  events.on('status', status );
  return true;
};
